include/albatross/src/evaluation/cross_validation_utils.hpp

```diff
@@ -162,6 +162,104 @@ leave_one_out_conditional(const JointDistribution &prior,
   return MarginalDistribution(loo_mean, loo_variance);
 }
 
+namespace details {
+
+// The following methods implement leave one group out cross validation
+// for more details see:
+//
+// https://swiftnav-albatross.readthedocs.io/en/latest/gp-details.html
+
+inline Eigen::VectorXd
+held_out_prediction(const Eigen::MatrixXd &inverse_block,
+                    const Eigen::VectorXd &y, const Eigen::VectorXd &v,
+                    PredictTypeIdentity<Eigen::VectorXd>) {
+  return y - inverse_block.ldlt().solve(v);
+}
+
+inline MarginalDistribution
+held_out_prediction(const Eigen::MatrixXd &inverse_block,
+                    const Eigen::VectorXd &y, const Eigen::VectorXd &v,
+                    PredictTypeIdentity<MarginalDistribution>) {
+  const auto A_ldlt = Eigen::SerializableLDLT(inverse_block);
+  const Eigen::VectorXd mean = y - A_ldlt.solve(v);
+  return MarginalDistribution(mean, A_ldlt.inverse_diagonal());
+}
+
+inline JointDistribution
+held_out_prediction(const Eigen::MatrixXd &inverse_block,
+                    const Eigen::VectorXd &y, const Eigen::VectorXd &v,
+                    PredictTypeIdentity<JointDistribution>) {
+  const auto A_inv = inverse_block.inverse();
```

Comment:
:O 

1) Can we form the mean by using `.ldlt().solve(v)`?

2) `.inverse()` uses the LU decomposition.  Since we know it's a sub-diagonal of the inverse of a covariance matrix, is there some way we can either not re-invert or preserve the triangle and permutation until we get here?

I would place less importance on this if I hadn't just spent several days trying to find numerically negative eigenvalues and asymmetry.

Reply:
Yeah we can do `.ldlt().solve(v)` ... but we need the actual inverse a few lines below.

Reply:
That actual inverse is the inverse of an inverse, right?

Reply:
I see, I'd have to think about that a bit more. Is the inverse of a block of the inverse just a block  of the original matrix (that's a mouthful)? I do see how it seems we could avoid a step, but I'd have to do some proof by python to know for sure.

Reply:
The inverse of a block of the inverse is definitely not just a block of the original matrix, sadly.  I just wondered whether there is a way to either avoid inverting again or at least use parts of the LDL decomp of the original matrix to form this inverse, rather than pivoting and LU decomposing from scratch.

Reply:
Any actions we want to take on this? If there's a suggestion for a simple change to improve things I'm happy to do it, but don't think refactoring the cross validation code to return decompositions is a good use of time at the moment.

Reply:
OK no problem.